tests/catch/stress/stream/streamEnqueue.cc

```diff
@@ -0,0 +1,227 @@
+  std::uniform_int_distribution<std::mt19937::result_type> genVal(0, maxVal);
+
+  auto enqueueKernelThread = [&](hipStream_t stream) {
+    auto iter = genWork(engine);  // Generate work to be done via thread
+    for (auto i = 0; i < iter; i++) {
+      auto index = genIndex(engine);  // Generate Index to add to
+      auto val = genVal(engine);      // Generate value to add to the destination
+      hostData[index].data += val;    // Replicate it on host
+      addVal<<<1, 1, 0, stream>>>(dPtr, static_cast<size_t>(index),
+                                  static_cast<unsigned long long>(val));  // And on device
+    }
+  };
+
+  hipStream_t stream{};
+  HIP_CHECK(hipStreamCreate(&stream));
+  std::vector<std::thread> threadPool{};
+  threadPool.reserve(hwThreads);
+
+  // Launch work
+  for (size_t i = 0; i < hwThreads; i++) {
+    threadPool.emplace_back(std::thread(enqueueKernelThread, stream));
+  }
+
+  // Wait for work to finish
+  for (auto& i : threadPool) {
+    i.join();
+  }
+
+  HIP_CHECK(hipStreamDestroy(stream));
+
+  auto hPtr = std::make_unique<unsigned long long[]>(hwThreads);
+  HIP_CHECK(
+      hipMemcpy(hPtr.get(), dPtr, sizeof(unsigned long long) * hwThreads, hipMemcpyDeviceToHost));
+
+  // Validate that CPU and GPU has the same results
+  for (size_t i = 0; i < hwThreads; i++) {
+    INFO("Check for Index " << i);
+    REQUIRE(hostData[i].data.load() == hPtr[i]);
+  }
+}
+
+__global__ void doOperation(int* dPtr, size_t size, int val) {
+  auto i = threadIdx.x;
+  atomicAdd(dPtr + i, val);
+}
+
+// Allocate mulitple stream for same device.
+// Same device stream operate on same memory
+TEST_CASE("Stress_StreamEnqueue_DifferentThreads_MultiGPU") {
+  int deviceCount{0};
+  HIP_CHECK(hipGetDeviceCount(&deviceCount));
+  REQUIRE(deviceCount > 0);
+
+  // Skip the test if devices less than 2
+  if (deviceCount <= 1) {
+    HIP_SKIP_TEST("Skipping because devices <= 1");
+    return;
+  }
+
+  constexpr size_t streamPerGPU{3};  // Stream per gpu
+
+  std::vector<hipStream_t> streamPool{};
+  streamPool.reserve(deviceCount * streamPerGPU);
+
+  std::map<hipStream_t, int*> streamToDeviceMemory;           // Map of stream and device memory
+  std::map<hipStream_t, AtomicWrap<int>> streamToHostMemory;  // Map of stream and host result
+  std::map<hipStream_t, size_t> streamToDeviceIndex;  // Map of stream and device it was created on
+
+  constexpr size_t size = 1024;
+
+  for (size_t i = 0; i < deviceCount; i++) {
+    HIP_CHECK(hipSetDevice(i));
+
+    for (size_t j = 0; j < streamPerGPU; j++) {
+      hipStream_t stream{nullptr};
+      HIP_CHECK(hipStreamCreate(&stream));
+      REQUIRE(stream != nullptr);
+      streamPool.push_back(stream);
+
+      int* dPtr{nullptr};
+      HIP_CHECK(hipMalloc(&dPtr, sizeof(int) * size));
+      REQUIRE(dPtr != nullptr);
+      HIP_CHECK(hipMemset(dPtr, 0, sizeof(int) * size));
+      HIP_CHECK(hipDeviceSynchronize());    // Wait till Memset is done
```

Comment:
hipMemset isn't async so you shouldn't need to do a device sync

Reply:
done